src/v/datalake/coordinator/coordinator.cc

```diff
@@ -47,11 +54,147 @@ std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
+            continue;
+        }
+        auto term_res = term_fut.get();
+        if (term_res.has_error()) {
+            vlog(
+              datalake_log.debug,
+              "Coordinator error while running in term {}: {}",
+              synced_term,
+              term_res.error());
+            continue;
+        }
+    }
+    co_return;
+}
+
+ss::future<checked<std::nullopt_t, coordinator::errc>>
+coordinator::run_until_term_change(model::term_id term) {
+    auto& raft = *stm_->raft();
+    vassert(!term_as_.has_value(), "Multiple calls to run_until_term_change");
+    ss::abort_source term_as;
+    term_as_ = term_as;
+    auto reset_term_as = ss::defer([this] { term_as_.reset(); });
+    vlog(datalake_log.debug, "Running coordinator loop in term {}", term);
+    while (raft.is_leader() && term == raft.term()) {
+        // Make a copy of the topics to reconcile, in case the map changes
+        // during this call.
+        // TODO: probably worth building a more robust scheduler.
+        chunked_vector<model::topic> topics;
+        for (const auto& [t, _] : stm_->state().topic_to_state) {
+            topics.emplace_back(t);
+        }
+        for (const auto& t : topics) {
+            // TODO: per topic means embarrassingly parallel.
+            auto commit_res
+              = co_await file_committer_.commit_topic_files_to_catalog(
+                t, stm_->state());
+            if (commit_res.has_error()) {
+                switch (commit_res.error()) {
+                case file_committer::errc::shutting_down:
+                    co_return errc::shutting_down;
+                case file_committer::errc::failed:
+                    vlog(
+                      datalake_log.debug,
+                      "Error while committing files to catalog for topic {}",
+                      t);
+                    continue;
+                }
+            }
+            // TODO: apply table retention periodically too.
+
+            auto updates = std::move(commit_res.value());
+            if (updates.empty()) {
+                // Nothing to replicate.
+                continue;
+            }
+            storage::record_batch_builder builder(
+              model::record_batch_type::datalake_coordinator, model::offset{0});
+            for (auto& update : updates) {
+                builder.add_raw_kv(
+                  serde::to_iobuf(mark_files_committed_update::key),
+                  serde::to_iobuf(std::move(update)));
+            }
+            auto repl_res = co_await stm_->replicate_and_wait(
+              term, std::move(builder).build(), as_);
+            if (repl_res.has_error()) {
+                auto e = convert_stm_errc(repl_res.error());
+                vlog(datalake_log.warn, "Replication failed {}", e);
+                co_return e;
+            }
+        }
+        auto sleep_res = co_await ss::coroutine::as_future(
+          ssx::sleep_abortable(commit_interval_(), as_, term_as));
```

Comment:
I think we may soon need a commit interval per topic at which point this logic may become a bit more complicated, this seems fine for now but curious if you already thought about.